src/main/java/com/google/summit/SummitAST.kt

```diff
@@ -0,0 +1,198 @@
+          else -> throw IllegalArgumentException("Unexpected file type")
+        },
+      charStream = CharStreams.fromPath(path),
+    )
+
+  /**
+   * Parses and translates the [string] and returns a [CompilationUnit] representing the AST if the
+   * operation was successful.
+   *
+   * The [compilation type][CompilationType] is determined by contents of [string].
+   */
+  fun parseAndTranslate(string: String) =
+    parseAndTranslate(
+      name = STRING_INPUT,
+      type = null,
+      charStream = CharStreams.fromString(string),
+    )
+
+  /**
+   * Parses and translates the [string] as [type] and returns a [CompilationUnit] representing the
+   * AST if the operation was successful.
+   */
+  fun parseAndTranslate(string: String, type: CompilationType) =
+    parseAndTranslate(
+      name = STRING_INPUT,
+      type = type,
+      charStream = CharStreams.fromString(string),
+    )
+
+  /**
+   * Parses and translates the [charStream] as [type] and returns a [CompilationUnit] representing the AST if
+   * the operation was successful.
+   *
+   * If [type] is not provided, the compilation type is determined by the contents of [charStream].
+   */
+  internal fun parseAndTranslate(
+    name: String,
+    type: CompilationType?,
+    charStream: CharStream
+  ): CompilationUnit? {
+    // Apex is a case-insensitive language and the grammar is
+    // defined to operate on fully lower-cased inputs.
+    val lexer = ApexLexer(CaseInsensitiveInputStream(charStream))
+    val tokens = CommonTokenStream(lexer)
+    val parser = ApexParser(tokens)
+
+    val errorCounter = SyntaxErrorListener()
+    lexer.addErrorListener(errorCounter)
+    parser.addErrorListener(errorCounter)
+
+    // Do parse as complete compilation unit
+    val tree =
+      when (type ?: determineCompilationType(charStream)) {
+        CompilationType.CLASS -> parser.compilationUnit()
+        CompilationType.TRIGGER -> parser.triggerUnit()
+      }
+
+    if (errorCounter.numErrors > 0) {
+      logger.atWarning().log("Failed to parse %s", name)
+      return null // failure
+    }
+
+    try {
+      val translator = Translate(name, tokens)
+      val ast = translator.translate(tree)
+      return ast
+    } catch (e: Translate.TranslationException) {
+      logger.atWarning().withCause(e).log("Failed to translate %s", name)
+      return null // failure
+    }
+  }
+
+  /**
+   * Determines the [CompilationType] of [charStream].
+   *
+   * If `class` or `trigger` is found before the body of the top-level declaration, the
+   * corresponding [CompilationType] is returned. Otherwise, [CompilationType.CLASS] is returned by
+   * default.
+   */
+  private fun determineCompilationType(charStream: CharStream): CompilationType {
+    fun findType(): CompilationType? {
+      val lexer = ApexLexer(CaseInsensitiveInputStream(charStream))
+
+      // Discard tokens inside body of declaration
+      val tokens = lexer.allTokens.takeWhile { it.type != ApexLexer.LBRACE }
```

Comment:
`Lexer.getAllTokens()` will lex the whole file and materialize the collection of all tokens. If you want to avoid that, I think you'll need to use a different method to get them one at a time.